src/ScaledPolynomialTransformFitter.cc

```diff
@@ -0,0 +1,465 @@
+        computeScaling(catalog, keys.output)
+    );
+}
+
+ScaledPolynomialTransformFitter::ScaledPolynomialTransformFitter(
+    afw::table::BaseCatalog const & data,
+    Keys const & keys,
+    int maxOrder,
+    double intrinsicScatter,
+    afw::geom::AffineTransform const & inputScaling,
+    afw::geom::AffineTransform const & outputScaling
+) :
+    _keys(keys),
+    _intrinsicScatter(intrinsicScatter),
+    _data(data),
+    _outputScaling(outputScaling),
+    _transform(
+        PolynomialTransform(maxOrder),
+        inputScaling,
+        outputScaling.invert()
+    ),
+    _vandermonde(data.size(), detail::computePackedSize(maxOrder))
+{
+    // Create a matrix that evaluates the max-order polynomials of all the (scaled) input positions;
+    // we'll extract subsets of this later when fitting to a subset of the matches and a lower order.
+    for (std::size_t i = 0; i < data.size(); ++i) {
+        afw::geom::Point2D input = getInputScaling()(_data[i].get(_keys.input));
+        // x[k] == pow(x, k), y[k] == pow(y, k)
+        detail::computePowers(_transform._poly._u, input.getX());
+        detail::computePowers(_transform._poly._v, input.getY());
+        // We pack coefficients in the following order:
+        // (0,0), (0,1), (1,0), (0,2), (1,1), (2,0)
+        // Note that this lets us choose the just first N(N+1)/2 columns to
+        // evaluate an Nth order polynomial, even if N < maxOrder.
+        for (int n = 0, j = 0; n <= maxOrder; ++n) {
+            for (int p = 0, q = n; p <= n; ++p, --q, ++j) {
+                _vandermonde(i, j) = _transform._poly._u[p] * _transform._poly._v[q];
+            }
+        }
+    }
+}
+
+void ScaledPolynomialTransformFitter::fit(int order) {
+    if (order < 0) {
+        order = _transform.getPoly().getOrder();
+    }
+    int const packedSize = detail::computePackedSize(order);
+    std::size_t nGood = 0;
+    if (_keys.valid.isValid()) {
+        for (auto const & record : _data) {
+            if (record.get(_keys.valid)) {
+                ++nGood;
+            }
+        }
+    } else {
+        nGood = _data.size();
+    }
+    // One block of the block-diagonal (2x2) unweighted design matrix M;
+    // m[i,j] = u_i^{p(j)} v_i^{q(j)}.  The two nonzero blocks are the same,
+    // because we're using the same polynomial basis for x and y.
+    Eigen::MatrixXd m = Eigen::MatrixXd::Zero(nGood, packedSize);
+    // vx, vy: (2x1) blocks of the unweighted data vector v
+    Eigen::VectorXd vx = Eigen::VectorXd::Zero(nGood);
+    Eigen::VectorXd vy = Eigen::VectorXd::Zero(nGood);
+    // sxx, syy, sxy: (2x2) blocks of the covariance matrix S, each of which is
+    // individually diagonal.
+    Eigen::ArrayXd sxx(nGood);
+    Eigen::ArrayXd syy(nGood);
+    Eigen::ArrayXd sxy(nGood);
+    Eigen::Matrix2d outS = _outputScaling.getLinear().getMatrix();
+    for (std::size_t i1 = 0, i2 = 0; i1 < _data.size(); ++i1) {
+        if (!_keys.valid.isValid() || _data[i1].get(_keys.valid)) {
```

Comment:
What is `_keys.valid`? `valid.isValid()` is a bit mind-bending Please document the Keys argument of the constructor (and all the other arguments) as requested in an earlier comment. Without that I have no idea what `_keys` is and what fields it contains.

Also, is `_keys.valid.isValid()` efficient to call? If in doubt, it might be better to set a local const variable and use that in the loop.